docs/cache_extensibility.md

```diff
@@ -0,0 +1,125 @@
+# MSAL Cache Key Extensibility Proposal
+
+This document looks at defining a strategy for extending how tokens are cached internally by MSAL, allowing app developers to cache multiple tokens by a custom key.
+
+## Functional Requirements
+
+- Allow higher level SDKs to extend the default cache key semantics that all MSALs come with.
+- Non-breaking changes
+- Implementations MUST be consistent across confidential client MSALs
+- `AcquireTokenForClient` (client_credentials) MUST support this mechanism. Other confidential client flows (web app, web api, ROPC) SHOULD support it.
+
+## Non-Functional Requirements
+
+- Performance of cache look-up operations does not degrade.
+- Forward compatibility strategy - older MSALs must continue to work on the with the same cache as the new ones, to support upgrade scenarios. If this is not possible, an intermediate version of MSAL must be released that supports this.
+- Logging must be extended so that MSAL developers can understand a cache miss.
+
+### Example scenarios that will use this extensibility point
+
+- Associate tokens with the client certificate used to obtain them.
+- Associtate tokens with the client secrets used to obtain them.
+- Allow an application to associate tokens with a SPIFEE type of identifier which is known upfront by the clients.
+
+### Prior art
+
+The cache key schema is defined [here](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/SSO/Schema.md) and extended for POP tokens [here](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/SSO/change_proposals/11232019-accesstoken_with_authscheme.md). A related cache extensiblity enhancement has been done [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/4922).
+
+## Developer Experience
+
+#### MSAL 
+
+```csharp
+
+var app = ConfidentialClientApplicationBuilder.Create("client_id")             
+              .WithClientCertifiacte(x509Cert)
+              .WithExperimentalFeatures(true)               // All extensiblity APIs remain experimental
+              .BuildConcrete();
+
+var result = await app.AcquireTokenForClient("https://graph.microsoft.com/.default")
+    .WithAdditionalCacheKeyComponents(new Dictionary<string, string>{ // New API
+        {"cert_thubprint", x509Cert.GetSha2THumbprint() }, 
+        {"spiffee_id", "37" }
+    });
+    .ExecuteAsync();   
+```
+
+After executing the AcquireToken instruction, MSAL shall associate the token with the existing components - authority, scope and client_id. In addition, it will assocaite the token with "cert_thumbprint" and "spifee_id". 
```

Comment:
"cert_thumbprint" and "spifee_id" are defined by the customer right?

Reply:
yes.